saw-central/src/SAWCentral/MRSolver/Term.hs

```diff
@@ -312,6 +313,22 @@ asLambdaName :: Recognizer Term LocalName
 asLambdaName (asLambda -> Just (nm, _, _)) = Just nm
 asLambdaName _ = Nothing
 
+----------------------------------------------------------------------
+-- * 'MonadTerm' type class
+----------------------------------------------------------------------
+
+-- | The class of monads that can build terms and substitute into them
+class Monad m => MonadTerm m where
+  mkTermF :: TermF Term -> m Term
+  liftTerm :: DeBruijnIndex -> DeBruijnIndex -> Term -> m Term
+  substTerm :: DeBruijnIndex -> [Term] -> Term -> m Term
+               -- ^ NOTE: the first term in the list is substituted for the most
+               -- recently-bound variable, i.e., deBruijn index 0
+
+instance (MonadTerm m, MonadTrans t, Monad (t m)) => MonadTerm (t m) where
+  mkTermF = lift . mkTermF
+  liftTerm n i t = lift $ liftTerm n i t
+  substTerm n s t = lift $ substTerm n s t
```

Comment:
I'm not a big fan of overlapping instances like this one... is it possible to remove it and just inline the (relatively short) method implementations in each instance that needs it? Alternatively, we could derive these `MonadTerm` instances using `DerivingVia` if we wanted to.

Reply:
It turns out that this instance was only used to call `traverseSubterms :: MonadTerm m => (Term -> m Term) -> Term -> m Term` at type `MaybeT MRM`. Because of the higher-order type of `traverseSubterms`, using `lift` doesn't work. But it was easy to replace the `MonadTrans` instance with a more specific one for `MaybeT` (7e4cad0c0c03ab9d7ee77fd0ee8d8e6440dc9815).